### PR TITLE
[doctool] Add markdown support && Static site generation

### PR DESCRIPTION
**Pull request by felixmulder**
This PR adds a bunch of fixes to the doc-tool as well as adding support for Markdown parsing and static site generation.

The end-goal of this PR is to be able to use Dottydoc to generate the complete website and documentation site for Dotty, but that's currently a ways off.

## List of changes / fixes ##
- [x] Fix non-static main method for dottydoc
- [x] Fix package structure, remove dead code warnings
- [x] Add doc-gen programs to aid development (requires a clone of dottydoc-client)
- [x] Add markdown parsing for comment bodies
- [x] Add fallback flag `-Xwiki-syntax`
- [x] Fix parsing of annotations like `@author` etc (i.e. either parsing by markdown or wiki syntax)
- [x] Fix summary parsing using the markdown AST
- [x] Decide upon static site generation strategy
    - emulating Jekyll style static site using Liquid and YAML-front matter
    - choosing bootstrap as default css
- [x] Setup structure for templates, `_layouts` etc
- [x] Fix recursive expansion of templates
- [x] Preserve page info in recursive expansion?
- [x] Fix `{% include template.html %}` construct in liqp parser
- [x] Implement main that can create site from specified dir
- [x] Copy folders and assets on gen
- [x] Implement `site.posts`
- [x] Port old site
- [x] Scala logo at top of toc
- [x] Generate API docs
- [x] Fix blogpost template (images and author)
- [x] Add annotations to tree
- [x] Delete old doc generation functionality from doc-tool
- [x] Fix reference links
- [x] Materialization of reference links causes terrible OOM errors, get to the bottom of it
- [x] Fix api generation for path dependent types
- [x] Get rid of "scala-native" logo (@densh)
- [x] Fix trailing html tags in comment generation
- [x] Remove implicitly added constructor vals
- [x] Implement rhs of type aliases
- [x] add sbt task `genDocs`
- [x] move `scala-logo.html` to `scala-logo.svg`, can include as svg, no problemo
- [x] link nested entities
- [x] fix `c.type` style return values
- [x] rewire `MemberLookup` to just return the path from root to found member
- [x] Merge package-objects with packages, **conclusion:** solved by moving comment and members to appropriate package and deleting `package object` node from AST.
- [x] Document `PackageObjectsPhase.scala`
- [x] Link companions (`Companion#hasCompanion`and `Companion#companionPath`)
- [x] Root package in `ctx.docbase.packages` is incomplete, some packages only exist in separate child packages in the same map. FIXME
- [x] Generalize `toc.html`
- [x] Referencing in static site `{% getDoc "scala.collection.Map" %}`
- [x] Add @OlivierBlanvillain to list of people working on dotty - I need a pic from you

## Feedback todo ##

- [x] Refactor comments
- [x] `case EmptyReference`
- [x] Scala default for naked code blocks
- [x] Check `inheritdoc` behaviour
- [x] Better reporting

### CODE REVIEW

File: doc-tool/src/dotty/tools/dottydoc/DocFrontEnd.scala [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+import dotc.typer.FrontEnd
+import dotc.core.Contexts.Context
+import dotc.CompilationUnit
+
+/** `DocFrontEnd` uses the Dotty `FrontEnd` without discarding the AnyVal
+  * interfaces for Boolean, Int, Char, Long, Byte etc.
+  *
+  * It currently still throws away Java sources by overriding
+  * `discardAfterTyper`.
```
Comment (smarter) [inline doc-tool/src/dotty/tools/dottydoc/DocFrontEnd.scala]: We should open an issue to support Java code like https://github.com/scala/scala/pull/5246

Comment (felixmulder) [inline doc-tool/src/dotty/tools/dottydoc/DocFrontEnd.scala]: Agreed, I'll have a look at the changes from the PR :)

**Comment by OlivierBlanvillain** [comment]
@felixmulder you can use this one: https://people.epfl.ch/olivier.blanvillain?lang=en

**Comment by felixmulder** [comment]
This is good enough for review now @OlivierBlanvillain.

File: doc-tool/test/MarkdownTests.scala [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+             |</ul>""".stripMargin, traitCmt)
+    }
+  }
+
+  @Test def handleCodeBlock = {
+    val source =
+      """
+      |package scala
+      |
+      |/** ```scala
```
Comment (OlivierBlanvillain) [inline doc-tool/test/MarkdownTests.scala]: I wonder is `scala` should be added as a default for "naked" code blocks

File: docs/docs/usage/dottydoc.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+Just as with Jekyll, the site is rendered in a `_site` directory.
+
+Using existing Templates and Layouts
+====================================
+Dottydoc uses the [Liquid](https://shopify.github.io/liquid/) templating engine
+and provides a number of custom filters and tags specific to Scala
+documentation.
+
+In Dottydoc, all templates can contain YAML front-matter. The front-matter
+is parsed an put into the `page` variable available in templates via Liquid.
```
Comment (OlivierBlanvillain) [inline docs/docs/usage/dottydoc.md]: s/an/and/

File: docs/docs/usage/dottydoc.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    - title: Usage
+      subsection:
+        - title: Dottydoc
+          url: docs/usage/dottydoc.html
+        - title: sbt-projects
+          url: docs/usage/sbt-projects.html
+```
+
+The `sidebar` key is mandatory, as well as `title` for each element. The
+default toc allows you to have subsections - albeit the current depth limit is
```
Comment (OlivierBlanvillain) [inline docs/docs/usage/dottydoc.md]: s/toc/table of content/

File: docs/docs/usage/dottydoc.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+Dottydoc denotes objects by ending their names in "$". To select `Object.range`
+you'd therefore write:
+
+```markdown
+[Object.range](scala.collection.immutable.List$.range)
+```
+
+Rendering Docstrings
+--------------------
+Sometimes you end up duplicating the docstring text in you documentation,
```
Comment (OlivierBlanvillain) [inline docs/docs/usage/dottydoc.md]: s/you/your/

File: docs/docs/usage/dottydoc.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+├── blog
+│   ├── _posts
+│   │   └── 2016-12-05-implicit-function-types.md
+│   └── index.md
+└── index.html
+```
+
+Sidebar
+=======
+Dottydoc provides you with the ability to create your own custom table of
```
Comment (OlivierBlanvillain) [inline docs/docs/usage/dottydoc.md]: s/provides you with/gives you/ (remove the repetition with the provide from the previous paragraph)

File: docs/docs/usage/dottydoc.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+│       └── 2016-12-05-implicit-function-types.md
+├── index.html
+└── sidebar.yml
+```
+
+Dottydoc Specific Tags and Behavior
+====================================
+Linking to API
+--------------
+If you for instance, want to link to `scala.collection.Seq` in a markdown file,
```
Comment (OlivierBlanvillain) [inline docs/docs/usage/dottydoc.md]: Is it intended not to have the `.immutable` here?

Comment (OlivierBlanvillain) [inline docs/docs/usage/dottydoc.md]: I can see from this PR that you love mutation, but still, :trollface:

File: docs/docs/usage/dottydoc.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+
+```markdown
+[Seq](scala.collection.immutable.Seq.isEmpty)
+```
+
+Dottydoc denotes objects by ending their names in "$". To select `Object.range`
+you'd therefore write:
+
+```markdown
+[Object.range](scala.collection.immutable.List$.range)
```
Comment (OlivierBlanvillain) [inline docs/docs/usage/dottydoc.md]: I've seen a different convention: `Seq.empty` for static member and `Seq#isEmpty` for non static.

File: sbt-bridge/src/xsbt/ScaladocInterface.scala [hunk 9]
```diff
@@ -18,9 +18,10 @@ class DottydocRunner(args: Array[String], log: Logger, delegate: xsbti.Reporter)
     val resources = getResources(args)
     val template  = getTemplate(resources)
 
-    template.fold(writeJson(index, outputFolder)) { tpl =>
-      buildDocs(outputFolder, tpl, resources, index)
-    }
+    // FIXME: temporarily disabled until new implementation in place
```
Comment (OlivierBlanvillain) [inline sbt-bridge/src/xsbt/ScaladocInterface.scala]: .

**Comment by felixmulder** [comment]
My brother @axelmulder (aka designer-guy 🍸 ), made this new logo for us (currently included in the PR):

screen shot 2017-01-31 at 17 34 08

Is this something that could get approval/feedback on? I set the color to something a bit milder than the usual firetruck red, but that's totally up for debate. I like purple 😈 

ping: @odersky, @lampepfl/dotty-team,  @heathermiller

**Comment by heathermiller** [comment]
@felixmulder NICE!!! It's cool without looking like a cheese grater or something. I like it 😄
(Though since Dotty is currently @odersky's baby, he may be more sensitive to Dotty logos than I)

In any case, very nice, @axelmulder 🍸!! (I assume the martini glass is obligatory when referring to @axelmulder 🍸) Thanks for taking the time to make a logo for us 😸

**Comment by felixmulder** [comment]
@OlivierBlanvillain - this PR is done for now. I'll merge tomorrow after I've talked to Fabien about generating the static site.

**Pull request merged**